File: leetcode/problems/13-roman-to-integer/13-roman-to-integer.go
package main

func romanToInt(s string) int {
	if len(s) < 0 {
		return 0
	}

	romanSymbols := map[byte]int{
		'I': 1,
		'V': 5,
		'X': 10,
		'L': 50,
		'C': 100,
		'D': 500,
		'M': 1000,
	}

	res, i, v1, v2 := 0, 0, 0, 0
	for i = 0; i < len(s)-1; i += 1 {
		v1 = romanSymbols[s[i]]
		v2 = romanSymbols[s[i+1]]
		if v2 > v1 {
			res += v2 - v1
			i += 1
		} else {
			res += v1
		}
	}

	if i == len(s) {
		return res
	}

	return res + romanSymbols[s[i]]
}

File: leetcode/problems/problem-specific/13.roman-to-integer/13.roman-to-integer.py
#
# @lc app=leetcode id=13 lang=python3
#
# [13] Roman to Integer
#

# @lc code=start

# * Hash Table Solution | O(n) Time | O(1) Space
# * n -> The length of s string

class Solution:
    def romanToInt(self, s: str) -> int:
        s_length = len(s)
        ROMAN_SYMBOLS = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }

        result = 0
        i = 0
        while i < s_length-1:
            cur_val, next_val = ROMAN_SYMBOLS[s[i]], ROMAN_SYMBOLS[s[i+1]]
            if next_val > cur_val:
                result += next_val - cur_val
                i += 1

            else:
                result += cur_val

            i += 1

        return result if i == s_length else result+ROMAN_SYMBOLS[s[i]]

# @lc code=end
